packages/tldraw/src/lib/shapes/text/RichTextArea.tsx

```diff
@@ -0,0 +1,242 @@
+		}
+	}
+
+	useEffect(() => {
+		// It's possible that by the time this hook runs, that a new shape is now being edited.
+		// Don't clear the text editor in that case.
+		if (!isEditing && editor.getEditingShapeId() === shapeId) {
+			editor.setEditingShapeTipTapTextEditor(null)
+		}
+	}, [editor, shapeId, isEditing])
+
+	useEffect(() => {
+		function selectAllIfEditing(event: { shapeId: TLShapeId }) {
+			if (event.shapeId === editor.getEditingShapeId()) {
+				setShouldSelectAllOnCreate(true)
+			}
+		}
+
+		function placeCaret(event: { shapeId: TLShapeId; point: { x: number; y: number } }) {
+			if (event.shapeId === editor.getEditingShapeId()) {
+				setInitialPositionOnCreate(event.point)
+			}
+		}
+
+		editor.on('select-all-text', selectAllIfEditing)
+		editor.on('place-caret', placeCaret)
+		return () => {
+			editor.off('select-all-text', selectAllIfEditing)
+			editor.off('place-caret', placeCaret)
+			setShouldSelectAllOnCreate(false)
+			setInitialPositionOnCreate(null)
+		}
+	}, [editor, isEditing])
+
+	const handleUpdate = (props: EditorEvents['update']) => {
+		handleChange({ richText: props.editor.state.doc.toJSON() })
+	}
+
+	const onKeyDown = (view: EditorView, event: KeyboardEvent) => {
+		if (event.key === 'Tab') {
+			handleTab(editor, view, event)
+		}
+
+		handleKeyDown(event)
+	}
+
+	if (!isEditing) return null
+	if (!tipTapConfig) return null
+
+	const { editorProps, ...restOfTipTapConfig } = tipTapConfig
+
+	return (
+		<div
+			id={tipTapId}
+			ref={ref}
+			tabIndex={-1}
+			data-testid="rich-text-area"
+			className="tl-rich-text tl-text tl-text-input"
+			onTouchEnd={stopEventPropagation}
+			onContextMenu={isEditing ? stopEventPropagation : undefined}
+			onPointerDownCapture={stopEventPropagation}
+			// On FF, there's a behavior where dragging a selection will grab that selection into
+			// the drag event. However, once the drag is over, and you select away from the textarea,
+			// starting a drag over the textarea will restart a selection drag instead of a shape drag.
+			// This prevents that default behavior in FF.
+			onDragStart={preventDefault}
+		>
+			<div className="tl-rich-text-tiptap">
+				<EditorProvider
+					autofocus
+					editable={isEditing}
+					onUpdate={handleUpdate}
+					onFocus={handleFocus}
+					onBlur={handleBlur}
+					onCreate={handleCreate}
+					immediatelyRender={true}
+					shouldRerenderOnTransaction={false}
+					editorProps={{
+						handleKeyDown: onKeyDown,
+						handleDoubleClick: (view, pos, event) => handleDoubleClick(event),
+						...editorProps,
+					}}
+					{...restOfTipTapConfig}
+					content={richText as JSONContent}
+				/>
+			</div>
+		</div>
+	)
+})
+
+// Prevent exiting the editor when hitting Tab.
+// Also, insert a tab character at the front of the line if the shift key isn't pressed,
+// otherwise if shift is pressed, remove a tab character from the front of the line.
+function handleTab(editor: Editor, view: EditorView, event: KeyboardEvent) {
```

Comment:
this looks fairly specific to the set of rich text things we've gone for by default in tldraw. could/should this be customizable? is there a way to supply something like this as part of the tiptap options we pass in?

Reply:
Ah, actually, it's quite different from the existing tldraw tab behavior.
This is to make the tab behavior work _in general_, as expected for list items.

I'm down to make an option to make it customizable but I think I'd wait for someone to ask for it, that can be easily added to the `textOptions` later as need be.